### PR TITLE
AI fixes

### PR DESCRIPTION
**Pull request by SpiritBaker**
Tweaks to Building order
New AI logic for giving subsidies
New AI logic for Edicts

### CODE REVIEW

File: common/ai_army/10_AI_forts.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+province = {
+    war = {
+        active = {
+            is_at_war = yes
+        }
+        eval_multiply = {
+            factor = 2
+			AND = {
+				owned_by = FROM
```
Comment (stackpoint) [inline common/ai_army/10_AI_forts.txt:9]: Is there any documentation on how this file is scoped?

**Comment by SpiritBaker** [comment]
"Small fix"
I tweaked building order a bit, to compensate for building cost from development.

File: common/state_edicts/02_promote_military_recruitment.txt [hunk 2]
```diff
@@ -36,100 +36,121 @@ edict_promote_military_recruitment = {
 	}
 	
 	modifier = {
-		local_manpower_modifier = 0.50
-		local_manpower = 0.250
+		local_manpower_modifier = 0.33
+		#local_manpower = 0.250
```
Comment (stackpoint) [inline common/state_edicts/02_promote_military_recruitment.txt:40]: Isn't reducing the percent bonus ontop of removing the flat bonus too much of a nerf?

File: common/state_edicts/02_promote_military_recruitment.txt [hunk 3]
```diff
@@ -36,100 +36,121 @@ edict_promote_military_recruitment = {
...
-		}
-		modifier = {
-			factor = 1.25
-			NOT = {
-				owner = {
-					manpower_percentage = 0.4
-				}
-			}
-		}
+		factor = 5
```
Comment (stackpoint) [inline common/state_edicts/02_promote_military_recruitment.txt:46]: The AI is still picking the edict with all of the modifier changes?

File: common/state_edicts/03_protect_trade.txt [hunk 4]
```diff
@@ -40,46 +40,80 @@ edict_protect_trade = {
 		province_trade_power_value = 1.0
 	}
 	color = {  113 11 43 }
+
 	ai_will_do = {
-	
-		# trade_income_percentage = 0.10
-	
-		factor = 30
+		factor = 5
```
Comment (stackpoint) [inline common/state_edicts/03_protect_trade.txt:45]: Was AI enabling protect trade too much?

File: common/ai_army/10_AI_forts.txt (hunk 1 above)
Comment (SpiritBaker) [reply common/ai_army/10_AI_forts.txt:9]: I've made it from example in vanilla folder

Comment (stackpoint) [reply common/ai_army/10_AI_forts.txt:9]: Are we sure that FROM scope is the enemy country? I couldn't find that in vanilla.

File: common/state_edicts/02_promote_military_recruitment.txt (hunk 3 above)
Comment (SpiritBaker) [reply common/state_edicts/02_promote_military_recruitment.txt:46]: AI will use it if there is a province with barracks of any tier, or a province with high base manpower.

File: common/state_edicts/03_protect_trade.txt (hunk 4 above)
Comment (SpiritBaker) [reply common/state_edicts/03_protect_trade.txt:45]: Older value was overwriting any economic consideration. And there was no modifier to stop AI from using edicts.

File: common/state_edicts/02_promote_military_recruitment.txt (hunk 2 above)
Comment (SpiritBaker) [reply common/state_edicts/02_promote_military_recruitment.txt:40]: This is vanilla value, and the edicts are still "cheaper" in a modpack.

File: common/ai_army/10_AI_forts.txt (hunk 1 above)
Comment (SpiritBaker) [reply common/ai_army/10_AI_forts.txt:9]: I think FROM is the owner in this case.

File: common/state_edicts/02_promote_military_recruitment.txt (hunk 2 above)
Comment (stackpoint) [reply common/state_edicts/02_promote_military_recruitment.txt:40]: Vanilla values are fine in this case?

File: common/ai_army/10_AI_forts.txt (hunk 1 above)
Comment (stackpoint) [reply common/ai_army/10_AI_forts.txt:9]: Owner of the army or of the province? Is there any way confirm this?

Comment (SpiritBaker) [reply common/ai_army/10_AI_forts.txt:9]: So here is developer comment

 Scripts in ai_army can be used to influence how the army AI behaves when certain conditions (Defined in the "active" trigger) are met.
It is intended not to control daily decision making of armies but to define a general attitude that will influence (or overwrite) hard-coded AI behaviour.
The "active" trigger (I.e. the trigger that decides what attitude is currently active) is evaluated when any of the following happens:
A war begins or ends
A country gains or loses a province
A year without wars or province changes passes

File: common/state_edicts/02_promote_military_recruitment.txt (hunk 2 above)
Comment (SpiritBaker) [reply common/state_edicts/02_promote_military_recruitment.txt:40]: This Edict was worth it even in vanilla. There is still editc lvls, so it gets even stronger for the same cost.

File: common/ai_army/10_AI_forts.txt (hunk 1 above)
Comment (stackpoint) [reply common/ai_army/10_AI_forts.txt:9]: My question is mostly ROOT vs FROM here.

Comment (SpiritBaker) [reply common/ai_army/10_AI_forts.txt:9]: Well the comments puts "Owner of the army" out of the question. I think it mainly increases  the value of the province, in a scoring system AI use.

Comment (stackpoint) [reply common/ai_army/10_AI_forts.txt:9]: Isn't "owned_by = FROM" redundant then if FROM is owner of province?

Comment (SpiritBaker) [reply common/ai_army/10_AI_forts.txt:9]: But the province in question can be owned by the enemy, this is a purely defensive thing.

**Review by stackpoint — APPROVED** [review]
Good to go

**Comment by stackpoint** [comment]
Whoops, spoke too soon:
```
[meantimetohappen.cpp:171]: unknown command '{' for MTTH in file common/ai_army/10_AI_forts.txt line : 8
[meantimetohappen.cpp:171]: unknown command 'FROM' for MTTH in file common/ai_army/10_AI_forts.txt line : 9
[meantimetohappen.cpp:171]: unknown command '{' for MTTH in file common/ai_army/10_AI_forts.txt line : 10
[meantimetohappen.cpp:171]: unknown command 'fort_15th' for MTTH in file common/ai_army/10_AI_forts.txt line : 11
[meantimetohappen.cpp:171]: unknown command 'fort_16th' for MTTH in file common/ai_army/10_AI_forts.txt line : 12
[meantimetohappen.cpp:171]: unknown command 'fort_17th' for MTTH in file common/ai_army/10_AI_forts.txt line : 13
[meantimetohappen.cpp:171]: unknown command 'fort_18th' for MTTH in file common/ai_army/10_AI_forts.txt line : 14
[ai_army_database.cpp:81]: Unsupported type army AI type }
```

**Comment by stackpoint** [comment]
Forgot to encase the conditions with "modifier = {    }"

**Comment by SpiritBaker** [comment]
Made a fix
[10_AI_forts.txt](https://github.com/TheDivexz/excelmakesbelhappypack/files/4646633/10_AI_forts.txt)

**Comment by stackpoint** [comment]
Looks good.

**Comment by stackpoint** [comment]
I'm seeing the AI bounce between fort defense and manpower edicts when not being sieged or having +90% manpower.

**Comment by SpiritBaker** [comment]
There is no changes to defense edict in this commit, but I can reduce the base Ai will do value to reduce the chance of this happening. Personally I've never seen this happening. 

About manpower edict, economic considerations work both ways. Why would AI disable the edict if the cost of running it is tiny percentage of his total income, or he have -100% to state maintenance.

**Comment by stackpoint** [comment]
If economic considerations isn't a concern then wouldn't running trade edict be better than running the MP edict while at +90% MP?

**Comment by SpiritBaker** [comment]
Why do think it's good idea to put % of total MP in a condition for the edict that increases total MP? 
It's not the micro human player can repeat, AI does that in one in game tic. 
Also I'm pretty sure AI factors total MP when he is trying to estimate his opponent strength.

**Comment by stackpoint** [comment]
If you want to extend the length of the edict to make it more like imperator to balance the AI's ability to micro then that's a completely different (valid) discussion. But looking at your 10,000 manpower pool and seeing you have 9990 manpower doesn't really seem like a difficult calculation for the human player to make. It's a difference of the ability to micro, but that shouldn't be a reason to deliberately affect the AI's ability to make a decision.

**Comment by SpiritBaker** [comment]
For edict with base manpower in it, the difference was very substantial. In a single tic AI had  the ability to double his manpower pull. 
With current lvl 1 edict it essentially half of that.

**Comment by stackpoint** [comment]
I definitely agree that flat manpower modifier was overpowered. I made a couple of tables on discord quantifying that.

**Pull request merged**